Allow renaming datasets & dataset with duplicate names

MichaelBuessemeyer opened the pull request:
### Further Notes:
- Quite some of the line changes are the result of moving the ObjectId class to the utils package so that all wk backend servers have access to this class.

### URL of deployed dev instance (used for testing):
- https://allowdatasetrenaming.webknossos.xyz

### Steps to test:
- Give two datasets the same name and check whether annotations and so on works
- Test whether the task system still works with duplicate dataset names
- check dataset upload
  - dataset upload
  - add remote
  - compose
- check moving datasets between folders (single & multiple at once)
- check whether the renaming works
- when there are duplicate names, the correct dataset should open upon view it (not another one with the same name)
- Test worker jobs (tested by @fm3)
- Test legacy routes
- And much much more :see_no_evil: 

### TODOs:
- [x] Add evolution and reversion
  - [x] testing needed
- [x] Test uploading: 
  - [x] Report upload fails
- [x] Adjust worker to newest job arguments as the dataset name can no longer be used to uniquely identify a dataset
- [x] rename `organization_name` in worker to `organization_id`. see #8038
- [x] Dataset Name settings field has an unwanted spinner (see upload view)
- [x] Check the job list
- [x] Properly implement legacy searching for datasets when old URI param is used 
- [x] Adjust legacy API routes to return dataset in old format
   - It is just an additional field. Thus, I would say it should be fine. 
- [x] datasets appear to be duplicated in the db 
  - Maybe these are created by jobs with an output dataset 
- [x] Fix dataset insert
- [x] Skeleton & VolumeTracings address a dataset via its name
  - Not really used only during task / annotation creation
  - Use heuristic upon upload and temporary patch the Tracing case classes to carry the datasetId during the creation process once the dataset is identified once.
  - Task creation works
  - [x] Needs testing
    - fix annotation upload 
  - [x] needs to support old nmls
- [x] Put datasetId into newly created nmls
- [x] In the backend `LinkedLayerIdentifier` still uses the datasetName as an identifier
  - used in wklibs, maybe just interpret the name as a path and work with this. in case it cannot be found the user needs to update wklibs. Add comment for this!
- ~~[ ] the dataset `C555_tps_demo` has quite some bucket loading errors. Unsure why some buckets do not work~~ The dataset seems to be broken. Could reproduce this on other branches
- [x] Notion-style URLs are missing (i.e. <name>-<id>, but only the id part is actually used)
- [x] Maybe remove `DatasetURIParser`
> I would also suggest to
> - [x] Execute the screenshot tests once for this branch. Take care to change the branch name in the CI config (or local command), so that the tests are actually ran on this branch.
- [x] Looks much better now :) Did you re-test that exceptions are caught as expected?
- [x] undo application.conf & snapshot changes -> in progress


### Issues:
- fixes #6613
- fixes #8038
- fixes #7560

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [x] Updated [documentation](../blob/master/docs) if applicable
- [x] Needs to execute a mirgation
- [x] Needs datastore update after deployment

MichaelBuessemeyer commented:
@fm3 The backend should be ready for a first glance :eyes:

fm3 reviewed (commented):
I started writing review comments, but that was now pre-empted by our in-person meeting :) still sending these two comments now.

MichaelBuessemeyer commented:
Note: We agree to renaming `dataset.path`  to `dataset.directory_name` as this is more clear. See here for more details: https://scm.slack.com/archives/C5AKLAV0B/p1729689141538769

hotzenklotz commented:
@coderabbitai full review

MichaelBuessemeyer commented:
@coderabbitai pause

as pr it tool large anyway

MichaelBuessemeyer commented:
I think this should be ready to go now. In case you notice some errors please tell me :pray:

fm3 commented on conf/webknossos.versioned.routes:
```diff
@@ -4,6 +4,7 @@
 # Note: keep this in sync with the reported version numbers in the utils.ApiVersioning trait
 
 # version log:
+ # changed in v9: Datasets are now identified by their id, not their name.
```
please add a slightly more verbose explanation for the concrete api changes here

fm3 commented on app/controllers/UserTokenController.scala:
```diff
@@ -113,7 +114,9 @@ class UserTokenController @Inject()(datasetDAO: DatasetDAO,
       for {
-        dataSourceBox <- datasetDAO.findOneByNameAndOrganization(dataSourceId.name, dataSourceId.team).futureBox
+        dataSourceBox <- datasetDAO
+          .findOneByDirectoryNameAndOrganization(dataSourceId.directoryName, dataSourceId.organizationId)
```
maybe a shortcut called findOneByDataSourceId could be created in the DAO, which does this expansion and forwards to findOneByDirectoryNameAndOrganization

fm3 reviewed (changes requested):
Really cool stuff! I had a look at the backend changes and added a couple of comments.
Most things are pretty small. Two bigger things I noticed
 - I’m not really sure about the reserveUpload protocol. It seems pretty involved, with identical fields being sent back and forth, with some values filled in remotely. Maybe we can simplify this, possibly using multiple requests
 - Looks like URIs with normalized name *and* id are not yet used? At least the OpenGraphService doesn’t seem to parse URIs in that way. Did you build that into the frontend? If not, did you talk to Norman about this? (I believe he had wished for this feature?)

I did not do any testing in this review round. I already tested a little earlier, and will do another round of testing in a later iteration.

MichaelBuessemeyer commented on frontend/javascripts/dashboard/dataset/dataset_settings_view.tsx:
```diff
@@ -367,14 +366,14 @@ class DatasetSettingsView extends React.PureComponent<PropsWithFormAndRouter, St
 
     if (dataset != null && this.didDatasourceChange(dataSource)) {
-      await updateDatasetDatasource(this.props.datasetId.name, dataset.dataStore.url, dataSource);
+      await updateDatasetDatasource(dataset.directoryName, dataset.dataStore.url, dataSource);
```
@daniel-wer I added this warning as it was kinda unexpected to me that the backend takes care of ignoring updates to the `datasource.id` field. Although this is how it has always been, do you think I should keep this info message or is this simply too much info for the user :)

@fm3 of cause I also value your opinion :)

MichaelBuessemeyer commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/uploading/ComposeService.scala:
```diff
@@ -50,7 +50,7 @@ object ComposeRequestLayer {
 }
 
-case class DataLayerId(name: String, owningOrganization: String)
+case class DataLayerId(path: String, owningOrganization: String)
```
I hope this should be cleared now :see_no_evil:

The `DataLayerId` was meant to uniquely identify the dataset of a layer which should be linked to the composed dataset. For consistency reasons I changed it to `dataSourceId: DataSourceId` in the ComposeRequestLayer. I hope that makes more sense now :D.

MichaelBuessemeyer replied:
Previously `DataLayerId` was designed to fulfull the exact same role as `DataSourceId` but with different field naming :woozy:. Moreover, the frontend did not send this info correctly. Now it should hopefully work :)

fm3 replied:
nice, much better!

MichaelBuessemeyer commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/uploading/ComposeService.scala:
```diff
@@ -64,32 +64,37 @@ class ComposeService @Inject()(dataSourceRepository: DataSourceRepository,
     for {
       dataSourceId <- Fox.successful(
-        DataSourceId(composeLayer.datasetId.name, composeLayer.datasetId.owningOrganization))
+        DataSourceId(composeLayer.datasetId.path, composeLayer.datasetId.owningOrganization))
```
see my comment above regardin `DataLayerId`

MichaelBuessemeyer reviewed (commented):
WIP applying @fm3 feedback. Submitting now to get a better overview of that's still an open TODO / uncommented

MichaelBuessemeyer commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/uploading/UploadService.scala:
```diff
@@ -51,23 +54,28 @@ object ReserveManualUploadInformation {
 
 case class LinkedLayerIdentifier(organizationId: Option[String],
                                  organizationName: Option[String],
+                                 // Filled by backend after identifying the dataset by name. Afterwards this updated value is stored in the redis database.
```
> Maybe let’s talk about this in person again.

Sure lets go :dango:

MichaelBuessemeyer commented on app/opengraph/OpenGraphService.scala:
```diff
@@ -91,33 +94,58 @@ class OpenGraphService @Inject()(datasetDAO: DatasetDAO,
-  private def datasetOpenGraphTagsWithOrganizationId(organizationId: String,
-                                                     datasetName: String,
-                                                     token: Option[String])(implicit ctx: DBAccessContext) =
+  private def getDatasetIdFromURIPath(datasetNameAndId: String): Option[ObjectId] = {
+    val idStrOpt = datasetNameAndId.split("-").lastOption
+    val idOpt = idStrOpt.flatMap(ObjectId.fromStringSync)
+    idOpt match {
+      case Some(validId) => Some(validId)
+      case None          => None
+    }
+  }
+
```
Here the id is extracted from the URL part

MichaelBuessemeyer commented on app/opengraph/OpenGraphService.scala:
```diff
@@ -91,33 +94,58 @@ class OpenGraphService @Inject()(datasetDAO: DatasetDAO,
-        datasetOpenGraphTagsWithOrganizationId(organizationId, datasetName, token)
-      case datasetRoute2Regex(organizationId, datasetName) =>
-        datasetOpenGraphTagsWithOrganizationId(organizationId, datasetName, token)
+      case datasetRoute1Regex(datasetIdStr) =>
+        val validDatasetIdOpt = getDatasetIdFromURIPath(datasetIdStr)
+        datasetOpenGraphTagsWithOrganizationId(validDatasetIdOpt, None, None, token)
+      case datasetRoute2Regex(datasetIdStr) =>
+        val validDatasetIdOpt = getDatasetIdFromURIPath(datasetIdStr)
+        datasetOpenGraphTagsWithOrganizationId(validDatasetIdOpt, None, None, token)
```
Here the opengraphtags are generated for routes that have the new URL schema

MichaelBuessemeyer reviewed (commented):
> Looks like URIs with normalized name and id are not yet used?

That is frontend only :+1: (except for the result link generation of jobs).

> At least the OpenGraphService doesn’t seem to parse URIs in that way. Did you build that into the frontend? If not, did you talk to Norman about this? (I believe he had wished for this feature?)

That should be implemented in the OpenGraphService. Please take a look at the newest comments

MichaelBuessemeyer commented:
Getting close to  +3k :tada: :laughing: 
![image](https://github.com/user-attachments/assets/bbbfcf59-1300-426c-b9af-3fea78ce4836)

fm3 commented:
Sorry I could not find all the right threads here to reply to, these are my latest notes on the open questions:

- TaskCreationParameters: let’s do the api adapter
- Job resultLink → let’s do 1 now and write a followup issue and discuss there if 2 or 3 is better
- messages → let’s go with the existing dataset.notFound
- NMLParser → catch exceptions in xml stuff, then use resulting box as input for the async stuff
- New wish: In settings view, add “Dataset Configuration as stored on datastore data-humerus at organization_name/datasetDirectoryName” to headline in advanced view
- In OpenGraphService, add comment what `datasetNameAndId.split("-").lastOption` is about (id+name uris)

MichaelBuessemeyer commented on frontend/javascripts/admin/task/task_create_form_view.tsx:
```diff
@@ -377,7 +376,10 @@ function TaskCreateFormView({ taskId, history }: Props) {
     if (taskId != null) {
       // either update an existing task
-      const newTask = { ..._.omit(formValues, "nmlFiles", "baseAnnotation"), boundingBox };
+      const newTask = {
+        ..._.omit(formValues, "nmlFiles", "baseAnnotation"),
+        boundingBox,
+      };
       const confirmedTask = await updateTask(taskId, newTask);
```
Here the frontend sends too much information to the backend. Afaik the value `newTask` contains more fields than what the backend expects. But this state has been working for quite some time. @fm3 @daniel-wer Do you think I should change that and omit all unexpected fields before sending the request or keep it as it has been in the past?

fm3 replied:
The backend generally ignores superfluous keys. I don’t have a strong opinion if the frontend should be cleaned up to send only the expected ones. Maybe Daniel has a preference :)

daniel-wer replied:
No strong opinion either. You can leave it as-is imo.

daniel-wer reviewed (commented):
First part of my frontend code review. Good stuff! I'll continue tomorrow :)

daniel-wer commented on frontend/javascripts/router.tsx:
```diff
@@ -603,46 +612,30 @@ class ReactRouter extends React.Component<Props> {
+                render={this.tracingViewModeLegacy}
               />
+              <Route path="/datasets/:datasetNameAndId/view" render={this.tracingViewMode} />
+              {/*maybe this also needs a legacy route?*/}
```
Yes I would think so, at least there are such links in the wild, for example in the [connectome viewer repository](https://github.com/scalableminds/connectome-viewer#:~:text=https%3A//webknossos.org/datasets/scalable_minds/l4dense_motta_et_al_demo_v2/sandbox/skeleton) and I've also sent such links via email to clients.

daniel-wer commented on frontend/javascripts/test/fixtures/dataset_server_object.ts:
```diff
@@ -1,6 +1,8 @@
+import type { APIDataset } from "types/api_flow_types";
 
-export default {
+const apiDataset: APIDataset = {
```
:pray: :tada:

daniel-wer commented on frontend/javascripts/test/model/model.spec.ts:
```diff
@@ -1,18 +1,22 @@
-// @ts-nocheck
```
Awesome, thank you!

daniel-wer commented on frontend/javascripts/test/puppeteer/dataset_rendering.screenshot.ts:
```diff
@@ -106,19 +106,38 @@ const datasetConfigOverrides: Record<string, PartialDatasetConfiguration> = {
 };
 
+const datasetNameToId: Record<string, string> = {};
+datasetNames.map(async (datasetName) => {
```
I think this test does not do what its description and assertion state. Instead it fetches the ids for all datasets. Doing that as part of a test seems hacky. Can't you use [test.before](https://github.com/avajs/ava/blob/main/docs/01-writing-tests.md#:~:text=or%20teardown%20code.-,test.before(),-registers%20a%20hook) for that?

daniel-wer commented on frontend/javascripts/test/snapshots/public-test/test-bundle/test/backend-snapshot-tests/misc.e2e.js.md:
```diff
@@ -77,12 +77,12 @@ Generated by [AVA](https://avajs.dev).
       isWkorgInstance: false,
-      jobsEnabled: false,
+      jobsEnabled: true,
       openIdConnectEnabled: false,
       optInTabs: [],
       publicDemoDatasetUrl: 'https://webknossos.org/datasets/scalable_minds/l4dense_motta_et_al_demo',
       recommendWkorgInstance: true,
       segmentAnythingEnabled: false,
       taskReopenAllowedInSeconds: 30,
-      voxelyticsEnabled: false,
+      voxelyticsEnabled: true,
```
Needs to be reverted before merging, together with the snap file.

daniel-wer commented on frontend/javascripts/types/api_flow_types.ts:
```diff
@@ -240,15 +241,15 @@ export type APIDatasetCompactWithoutStatusAndLayerNames = Pick<
 >;
 export type APIDatasetCompact = APIDatasetCompactWithoutStatusAndLayerNames & {
-  id?: string;
+  id: string; // Open question: Why was this optional?, The backend code clearly always returns an id ... :thinking:
```
Could be legacy or an oversight, but I don't know. I think it's fine to change if typescript doesn't complain.

daniel-wer reviewed (commented):
Second half of my PR review :) I haven't done any testing yet. Would you say now is a good time to do so?

I would also suggest to
- [ ] Execute the screenshot tests once for this branch. Take care to change the branch name in the CI config (or local command), so that the tests are actually ran on this branch.

MichaelBuessemeyer reviewed (commented):
Thanks for the feedback @daniel-wer. I applied / commented each of your comments. Please have another look. Testing should hopefully work :)

fm3 commented on app/models/annotation/nml/NmlParser.scala:
```diff
@@ -41,12 +47,14 @@ object NmlParser extends LazyLogging with ProtoGeometryImplicits with ColorGener
+            basePath: Option[String] = None)(implicit m: MessagesProvider,
+                                             ec: ExecutionContext,
+                                             ctx: DBAccessContext): Fox[NmlParseSuccessWithoutFile] = {
+    val foxInABox = try {
```
Looks much better now :) Did you re-test that exceptions are caught as expected?

MichaelBuessemeyer replied:
> Looks much better now :)

Thanks

> Did you re-test that exceptions are caught as expected?

I put it on the todo list :)

fm3 commented on app/models/dataset/DatasetService.scala:
```diff
@@ -164,16 +168,17 @@ class DatasetService @Inject()(organizationDAO: OrganizationDAO,
       case _ =>
-        insertNewDataset(dataSource, dataStore).toFox.map(Some(_))
+        insertNewDataset(dataSource, dataSource.id.directoryName, dataStore).toFox
+          .map(Some(_)) // TODO: Discuss how to better handle this case
```
Hmm, I see. Yes, it’s kind of unexpected. Unfortunately I don’t really have a good idea here. Since this was not introduced in this PR, I’d leave it as is for the time being.

fm3 reviewed (commented):
Getting close! I added a few more comments/replied to threads.

I think the two open backend topics are the dataset upload protocol and the ParamFailure thing for NML upload. I hope I’ll have time for both on monday.

daniel-wer reviewed (commented):
I was only able to do a quick test, but will continue testing on Monday :) I tested with the currently deployed dev instance, not sure whether that used the most recent code.

- [x] When clicking on the dataset name link at the top of the dataset settings page (e.g. https://allowdatasetrenaming.webknossos.xyz/datasets/l4_sample-6735ed6c010000d2004849e6/edit), the link https://allowdatasetrenaming.webknossos.xyz/datasets/6735ed6c010000d2004849e6 which makes webknossos crash

MichaelBuessemeyer commented on app/models/dataset/WKRemoteDataStoreClient.scala:
```diff
@@ -68,7 +68,7 @@ class WKRemoteDataStoreClient(dataStore: DataStore, rpc: RPC) extends LazyLoggin
 
   def findPositionWithData(dataset: Dataset, dataLayerName: String): Fox[JsObject] =
-    rpc(s"${dataStore.url}/data/datasets/${dataset._id}/layers/$dataLayerName/findData")
+    rpc(s"${dataStore.url}/data/datasets/${dataset._organization}/${dataset.directoryName}/layers/$dataLayerName/findData")
```
@frcroth thanks for adding the dataset upload test. This already saved me twice in this pr :see_no_evil: 

=> one place is here, where a request to the dataset was accidentally based on the dataset id and not on orga and the new directoryName

MichaelBuessemeyer commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/uploading/UploadService.scala:
```diff
@@ -31,8 +31,8 @@ import scala.concurrent.ExecutionContext
+    newDatasetId: Option[String],
     organization: String,
     totalFileCount: Long,
     filePaths: Option[List[String]],
```
And the other occurence is here, where I accidentally made the new upload info a requirement which would have crashed uploads tried with wklibs

MichaelBuessemeyer commented on frontend/javascripts/router.tsx:
```diff
@@ -456,7 +491,23 @@ class ReactRouter extends React.Component<Props> {
+                  const { datasetId, datasetName } = getDatasetIdOrNameFromReadableURLPart(
+                    match.params.datasetNameAndId,
+                  );
+                  if (datasetName) {
```
Kinda yes: There are links like `/datasets/:datasetName/view` out in the wild. Very old links. We even have a special backend route to get the organization id for a dataset base only on its name to resolve such old links. See: `getOrganizationForDataset`. 

The problem I am trying to solve here is that a routes ``/datasets/:datasetName/view` and ` `/datasets/:datasetId/view` have the same "matching string" and thus need to be decided by programmatically. I modified `getDatasetIdOrNameFromReadableURLPart` to make some guesswork on whether the passed argument for the match (`match.params.datasetNameAndId`)  contains an actual datasetId or just a name. In case it is a name I fall back to treating the legacy link and trying to resolve it to the newest schema.

As this is necessary for view links. there might also be the same problem for `.../edit` routes. I think it is very unlikely that there are some edit links with there very old URL schema out there. But in case they are out there, this code here takes care of it. I'd say this is not needed but the new return type of `getDatasetIdOrNameFromReadableURLPart` forces the user to consider legacy links. And I think this is a good thing so it won't be forgotten in the future. Therefore, I'd also keep the handling of legacy `/edit` links, bc it feels more case complete to me. But feel free to argue.

Besides that, handling the very old `/view` links is a must :)

daniel-wer replied:
Thank you for elaborating. Please mention this in a comment as this kind of knowledge gets easily lost otherwise :)

daniel-wer reviewed (commented):
This works really well! My few notes from more extensive testing:

- [ ] In the datasource-properties there is a field like
  ```
  "id": {
    "name": "asdasdasd",
    "team": "sample_organization"
  }
    ```
    which is not changed if the dataset name is changed. Is this intentional?
- [ ] Moving a single dataset from the dataset list into a folder using drag&drop fails with `Could not move dataset. Please try again.`. However, the batch moving of multiple datasets works.
- [ ] I could not test the jobs. Did you do that locally?

I tested what I could think of, but this change affects many areas of wk, so it would be great if @fm3 could test as well :) Because it's a little more involved I'm mentioning that there's no need to test the dataset composition feature using NMLs again - I did that already and it worked well.

fm3 commented:
> In the datasource-properties there is a field [`"id"`] which is not changed if the dataset name is changed. Is this intentional?

This is ok from my side. The backend does not use this field. Instead, the location on disk/datastore filesystem counts. It’s there only for backwards compatibility. Maybe we should set it to emptystring in the future to avoid confusion. But that does not need to be part of this PR. @markbader do you know if the libs use the `id` field of the datasource-properties json for anything?

> it would be great if @fm3 could test as well

Will do! I’ll have a closer look at the worker jobs then (also compare vx pr https://github.com/scalableminds/voxelytics/pull/3743 )

markbader commented:
> do you know if the libs use the id field of the datasource-properties json for anything?

@fm3 The libs currently do not consider the path name but use the `id` field of the datasource-properties.json file for accessing the name of the dataset. So the getter and setter of `Dataset.name` use it. I don't see any other occurences where the `id` field is used.

fm3 commented:
> The libs currently do not consider the path name but use the id field of the datasource-properties.json file for accessing the name of the dataset. So the getter and setter of Dataset.name use it. I don't see any other occurences where the id field is used.

Hmm ok I guess we’ll have to change that at some point when adapting to this renamable datasets stuff. But shouldn’t break anything now :crossed_fingers:

MichaelBuessemeyer commented on frontend/javascripts/dashboard/dataset/dataset_settings_data_tab.tsx:
```diff
@@ -184,6 +184,8 @@ function SimpleDatasetForm({
             <Row gutter={48}>
               <Col span={24} xl={12}>
                 <FormItemWithInfo
+                  // The dataset name is not synced with the datasource.id.name in the advanced settings, because datasource.id represents a DataSourceId
+                  // where datasource.id.name represents the dataset's directoryName and not the dataset's name.
```
@daniel-wer I added a comment to the dataset name form item explaining why this is not in sync with `datasource.id.name`

MichaelBuessemeyer commented:
>  Moving a single dataset from the dataset list into a folder using drag&drop fails with Could not move dataset. Please try again.. However, the batch moving of multiple datasets works.

Should be fixed now. Was a little buggy :bug:. While fixing this, I tired to also improve the typing

fm3 reviewed (approved):
Backend LGTM :)

I took the liberty to push two commits, one renames some local variables, and one fixes starting an infer_with_model job in the AiModelController. Hope that’s ok for you!

I also did some testing in addition to Daniel’s
- [x] Some dashboard interactions
- [x] Dataset upload
- [x] Some annotation interactions
- [x] Some dataset settings view interactions
- [x] Worker jobs
  - [x] infer mitos
  - [x] infer neurons
  - [x] train model
  - [x] infer with model
    → I fixed a line in AiModelController, now it works
  - [x] render animation
  - [x] compute meshfile
  - [x] compute_segment_index_file
  - [x] convert_to_wkw
  - [ ] export_tiff (broken on master too, but looks ok)
  - [x] align_sections
  - [x] materialize_volume_annotation
  - [x] find_largest_segment_id
- [x] SQL migration
  → I got a “there are 64 updated datasets” toast in the dashboard afterwards. Maybe suppress that if that is easy. If not, it’s also ok as is.

Please add a changelog entry and clean up the PR description. Then, with a green CI, this is ready to go from my side!

MichaelBuessemeyer commented:
> I took the liberty to push two commits, one renames some local variables, and one fixes starting an infer_with_model job in the AiModelController. Hope that’s ok for you!

Sure, thanks a lot :pray:

MichaelBuessemeyer commented on frontend/javascripts/dashboard/dataset/queries.tsx:
```diff
@@ -490,20 +485,22 @@ function diffDatasets(
   const changedDatasets = both
-    .map((name) => newDatasetsDict[name])
+    .map((id) => newDatasetsDict[id])
     .filter((newDataset) => {
-      const oldDataset = oldDatasetsDict[newDataset.name];
-      return !_.isEqualWith(oldDataset, newDataset, (_objValue, _otherValue, key) => {
-        if (key === "lastUsedByUser") {
-          // Ignore the lastUsedByUser timestamp when diffing datasets.
+      const oldDataset = oldDatasetsDict[newDataset.id];
+      return !_.isEqualWith(oldDataset, newDataset, (oldValue, newValue, key) => {
+        const didUpgradeToRenamableDS =
+          !("directoryName" in oldValue) && "directoryName" in newValue; // TODO: Can be remove after a few weeks / months.
+        if (key === "lastUsedByUser" || didUpgradeToRenamableDS) {
+          // Ignore the lastUsedByUser timestamp when diffing datasets and migrating datasets to new renamable version.
           return true;
```
> → I got a “there are 64 updated datasets” toast in the dashboard afterwards. Maybe suppress that if that is easy. If not, it’s also ok as is.

I could not reproduce this toast :see_no_evil: (maybe I am blind).
But I tried to remove the toast anyway by hiding my dataset

@daniel-wer I included a todo to remove this check after some time, when most users should have at least once opened the dashboard and got the dataset in the updated version. What do you think about this "fix" by ignoring an updated dataset based on whether it was migrated to the new data schema?

=> This might hide datasets that have updated since the last visit, as they also got the `directoryName` changed. Do you think, that's ok?

daniel-wer replied:
> => This might hide datasets that have updated since the last visit, as they also got the directoryName changed. Do you think, that's ok?

I think that's totally fair. This is more of a convenience feature.

MichaelBuessemeyer commented on app/models/task/TaskService.scala:
```diff
@@ -45,6 +47,8 @@ class TaskService @Inject()(conf: WkConf,
         "team" -> team.name,
         "type" -> taskTypeJs,
         "dataSet" -> dataset.name,
+        "datasetName" -> dataset.name,
+        "datasetId" -> dataset._id, // Only used for csv serialization in frontend.
```
here

MichaelBuessemeyer commented on frontend/javascripts/test/snapshots/public-test/test-bundle/test/backend-snapshot-tests/annotations.e2e.js.md:
```diff
@@ -461,6 +464,8 @@ Generated by [AVA](https://avajs.dev).
         created: 'created',
         creationInfo: null,
         dataSet: '2012-06-28_Cortex',
+        datasetId: '570b9f4e4bb848d0885ee711',
+        datasetName: '2012-06-28_Cortex',
         editPosition: [
```
@fm3 I just noticed by check which files where updated. I thought we wanted to avoid having `datasetName` and `dataSet` in parallel? 

Although checking: https://github.com/scalableminds/webknossos/pull/8075#discussion_r1827721741 shows that we did not clearly have this decision. Should I include a legacy route for anything that returns / receives a task? There is already a legacy route for creating and updating a task, but the publicWrites in `TaskService` still returns 
```
 "dataSet" -> dataset.name,
 "datasetName" -> dataset.name,
 "datasetId" -> dataset._id, // Only used for csv serialization in frontend.
```
Should I add a legacy route for this to not always send dataSet and datasetName?

MichaelBuessemeyer replied:
This can also be seen in `tasks.e2e.js.md`

daniel-wer commented on CHANGELOG.unreleased.md:
```diff
@@ -15,6 +15,7 @@ For upgrade instructions, please check the [migration guide](MIGRATIONS.released
 
 ### Changed
 - Reading image files on datastore filesystem is now done asynchronously. [#8126](https://github.com/scalableminds/webknossos/pull/8126)
+- Dataset can now be renamed and can have duplicate names. [#8075](https://github.com/scalableminds/webknossos/pull/8075)
```
```suggestion
- Datasets can now be renamed and can have duplicate names. [#8075](https://github.com/scalableminds/webknossos/pull/8075)
```

daniel-wer reviewed (approved):
LGTM assuming the screenshot tests were successful at some point :+1:

MichaelBuessemeyer commented:
I now removed the `dataSet` field in jsonified task object and added the legacy adaption. 
I checked routes that needed adaptation and came up with that the following do not need adaptation or do they?
```
- /tasks/:id/assign                                     controllers.TaskController.assignOne
- /taskTypes/:id/tasks                                  controllers.TaskController.listTasksForType
- /tasks/list                                           controllers.TaskController.listTasks
- /user/tasks/peek                                      controllers.TaskController.peekNext
- /tasks/:id                                            controllers.TaskController.update
- /annotations/:id/addAnnotationLayer                   controllers.AnnotationController.addAnnotationLayerWithoutType
- /annotations/:typ/:id/addAnnotationLayer              controllers.AnnotationController.addAnnotationLayer
- /datasets/:datasetId/createExplorational              controllers.AnnotationController.createExplorational
- /annotations/:id/downsample                           controllers.AnnotationController.downsampleWithoutType
- /annotations/:typ/:id/downsample                      controllers.AnnotationController.downsample
- /annotations/:typ/:id/duplicate                       controllers.AnnotationController.duplicate
- /annotations/:typ/:id/editLockedState                 controllers.AnnotationController.editLockedState
- /annotations/:typ/:id/finish                          controllers.AnnotationController.finish
- /datasets/:datasetId/sandbox/:typ                     controllers.AnnotationController.getSandbox
- /annotations/:typ/:id/info                            controllers.AnnotationController.info
- /annotations/:typ/:id/makeHybrid                      controllers.AnnotationController.makeHybrid
- /annotations/:typ/:id/merge/:mergedTyp/:mergedId      controllers.AnnotationController.merge
- /annotations/:typ/:id/reopen                          controllers.AnnotationController.reopen
- /annotations/:typ/:id/reset                           controllers.AnnotationController.reset
- /annotations/:typ/:id/transfer                        controllers.AnnotationController.transfer
```

I'll do testing of the new legacy routes later

MichaelBuessemeyer commented:
Hi @fm3,

I added the required legacy routes to remove `dataSet` from the json serialized task objects in the newest API version. Please find the following links & curl script to test these routes with the dev instance. All ids included & dataset names should exist on the dev instance and therefore just clicking the links should work. As these are legacy routes, the results should include the legacy field `dataSet` in each task object. Some routes return an annotation which itself has a task.

For the curl commands, you first need to fill in the id cookie before you can run the scripts.

Before testing here is one more important thing to double check: The legacy routes now include both: `dataSet` and `datasetName`. Can wklibs handle / ignore new unexpected json fields?

#### Checklist for testing new Legacy Routes
- https://allowdatasetrenaming.webknossos.xyz/api/v8/tasks/67464831010000ea00ca5839
- https://allowdatasetrenaming.webknossos.xyz/api/v7/tasks/67464831010000ea00ca5839
- https://allowdatasetrenaming.webknossos.xyz/api/v6/tasks/67464831010000ea00ca5839
- https://allowdatasetrenaming.webknossos.xyz/api/v5/tasks/67464831010000ea00ca5839
- https://allowdatasetrenaming.webknossos.xyz/api/v4/tasks/67464831010000ea00ca5839 -> should show error
- https://allowdatasetrenaming.webknossos.xyz/api/v8/projects/6735ed2e0100008c00484975/tasks
- https://allowdatasetrenaming.webknossos.xyz/api/v7/projects/6735ed2e0100008c00484975/tasks
- ...
- https://allowdatasetrenaming.webknossos.xyz/api/v8/annotations/6746db83010000b100416ed2/info?timestamp=0
- https://allowdatasetrenaming.webknossos.xyz/api/v7/annotations/6746db83010000b100416ed2/info?timestamp=0
- ...
- https://allowdatasetrenaming.webknossos.xyz/api/v8/tasks/6746482e010000d500ca5835/annotations
- https://allowdatasetrenaming.webknossos.xyz/api/v7/tasks/6746482e010000d500ca5835/annotations
- ...

Testing create & update task -> :shurg: I did this manually in firefox with the edit & resent feature on the dev instance.

### In case you want to test more complicated routes:

#### curl command for checking task creation (setting id cookie is required)
```bash
curl 'https://allowdatasetrenaming.webknossos.xyz/api/tasks' \
  -H 'accept: application/json' \
  -H 'accept-language: en-US,en;q=0.9,de;q=0.8' \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/json' \
  -H 'cookie: id=<fill-me-in>' \
  -H 'origin: https://allowdatasetrenaming.webknossos.xyz' \
  -H 'pragma: no-cache' \
  -H 'priority: u=1, i' \
  -H 'referer: https://allowdatasetrenaming.webknossos.xyz/tasks/create' \
  -H 'sec-ch-ua: "Google Chrome";v="131", "Chromium";v="131", "Not_A Brand";v="24"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Linux"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-origin' \
  -H 'user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36' \
  --data-raw '[{"taskTypeId":"63721e2cef0100470266c485","neededExperience":{"domain":"sampleExp","value":1},"pendingInstances":1,"projectName":"sampleProject","boundingBox":null,"dataSet":"kiwi","editPosition":[0,0,0],"editRotation":[0,0,0],"baseAnnotation":null}]'
```

#### Test task update: (setting id cookie is required)
```bash
curl 'https://allowdatasetrenaming.webknossos.xyz/api/tasks/6746ecc6010000d80015ea8c' \
  -X 'PUT' \
  -H 'accept: application/json' \
  -H 'accept-language: en-US,en;q=0.9,de;q=0.8' \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/json' \
  -H 'cookie: id=<fill-me-in>' \
  -H 'origin: https://allowdatasetrenaming.webknossos.xyz' \
  -H 'pragma: no-cache' \
  -H 'priority: u=1, i' \
  -H 'referer: https://allowdatasetrenaming.webknossos.xyz/tasks/6746ecc6010000d80015ea8c/edit' \
  -H 'sec-ch-ua: "Google Chrome";v="131", "Chromium";v="131", "Not_A Brand";v="24"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Linux"' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-origin' \
  -H 'user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36' \
  --data-raw '{"taskTypeId":"63721e2cef0100470266c485","neededExperience":{"domain":"sampleExp","value":1},"pendingInstances":2,"projectName":"sampleProject","boundingBox":null,"datasetId":"kiwi","editPosition":[0,0,0],"editRotation":[0,0,0]}'
```

fm3 commented on app/controllers/LegacyApiController.scala:
```diff
@@ -30,18 +29,139 @@ object LegacyCreateExplorationalParameters {
                                     sil: Silhouette[WkEnv])(implicit ec: ExecutionContext, bodyParsers: PlayBodyParsers)
     extends Controller {
 
+  /* to provide v8, remove legacy routes */
```
Not sure I understand this comment. What legacy routes are removed? (I mean, the PR removes some, but this comment appears to be about the following code lines). Maybe it should be something like “to provide v8, add dataSet to task json?” Or maybe remove these comments altogether.

MichaelBuessemeyer replied:
> Or maybe remove these comments altogether.

Something like these comments were present previously (I think). Thus, I kept them to "order" the code a little :)

I can also remove them in case you prefer that :)

fm3 commented on CHANGELOG.unreleased.md:
```diff
@@ -26,6 +27,7 @@ For upgrade instructions, please check the [migration guide](MIGRATIONS.released
 - Fix a bug where dataset uploads would fail if the organization directory on disk is missing. [#8230](https://github.com/scalableminds/webknossos/pull/8230)
 
 ### Removed
+- Removed legacy routes for versions 2,3 and 4. [#8075](https://github.com/scalableminds/webknossos/pull/8075)
```
```suggestion
- Removed support for HTTP API versions 3 and 4. [#8075](https://github.com/scalableminds/webknossos/pull/8075)
```
Could you also mention this in the migration guide please? (If someone has old code interacting with the HTTP API, they need to update)

fm3 reviewed (approved):
LGTM & Works for me =) Thanks! I added two more small comments. Also, have a look at the merge conflicts. None of that should be much work. Let’s ship it! :shipit:

MichaelBuessemeyer commented:
merge conflicts & your two comments are done now. So it's ready for :ship:

fm3 commented:
:tada:
